src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/NuGetProjectService.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.PackageManagement.VisualStudio;
+using NuGet.Packaging;
+using NuGet.ProjectManagement;
+using NuGet.ProjectManagement.Projects;
+using NuGet.VisualStudio.Contracts;
+
+namespace NuGet.VisualStudio.Implementation.Extensibility
+{
+    internal class NuGetProjectService : INuGetProjectService
+    {
+        private readonly IVsSolutionManager _solutionManager;
+
+        public NuGetProjectService(IVsSolutionManager solutionManager)
+        {
+            _solutionManager = solutionManager ?? throw new ArgumentNullException(nameof(solutionManager));
+        }
+
+        public async Task<InstalledPackagesResult> GetInstalledPackagesAsync(Guid projectId, CancellationToken cancellationToken)
+        {
+            // Just in case we're on the UI thread, switch to background thread. Very low cost (does not schedule new task) if already on background thread.
+            await TaskScheduler.Default;
+
+            NuGetProject project = await _solutionManager.GetNuGetProjectAsync(projectId.ToString());
+            if (project == null)
+            {
+                return NuGetContractsFactory.CreateInstalledPackagesResult(InstalledPackageResultStatus.ProjectNotReady, packages: null);
```

Comment:
What if the guid is not for a valid project? 
I think ProjectNotReady status is acceptable, jsut wanted to make sure you've considered it.

Reply:
It's one of the reasons I listed in the remarks for `ProjectNotReady`'s XMLDoc.

I re-used `ProjectNotReady`, rather than creating an `UnknownProject` for two reasons. The primary reason is because I knew how the service was going to be implemented, using `ProjectSystemCache`, and knowing how that cache is populated, we don't have information about whether a project guid is not in the solution, is in the solution but is unloaded, is in the solution, but a project type that doesn't support nuget, or is a CPS project that we haven't got a nomination for yet. So, knowing the implementation details, I designed the API contract to allow myself to be lazy. Not a good reason.  The second reason is because I didn't think callers of the API would be able to use the information. Why would a component call this API if the project guid didn't exist in the solution, or is unloaded? How would getting `InvalidGuid` instead of `ProjectNotReady` help them?  I didn't think it was worthwhile.

Reply:
Yeah, ofc :D 
Makes sense. You considered it and have a good reasoning for it, and I think it makes sense.